File: chapter4/governance-with-scripts/tags-corrector/main.py
import boto3

def lambda_handler(event, context):
    ec2 = boto3.client('ec2')
    
    response = ec2.describe_instances() #A
    
    for reservation in response['Reservations']:
        for instance in reservation['Instances']:
            instance_id = instance['InstanceId']
            tags = instance.get('Tags', [])
            
            corrected_tags = []
            for tag in tags:
                key = tag['Key']
                value = tag['Value']
                
                corrected_key = key.capitalize() #B
                corrected_value = value.capitalize()
                
                if key != corrected_key or value != corrected_value:
                    corrected_tags.append({
                        'Key': corrected_key,
                        'Value': corrected_value
                    })
            
            if corrected_tags: #C
                ec2.create_tags(
                    Resources=[instance_id],
                    Tags=corrected_tags
                )
                print(f"Corrected tags for instance {instance_id}")

    return {
        'statusCode': 200,
        'body': 'Tag correction completed'
    }

File: chapter4/governance-with-scripts/tags-corrector/main.go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
)

func handler() (map[string]interface{}, error) {
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		log.Fatalf("Unable to load SDK config, %v", err)
	}

	svc := ec2.NewFromConfig(cfg)

	resp, err := svc.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
	if err != nil {
		log.Fatalf("Failed to describe instances: %v", err)
	}

	for _, reservation := range resp.Reservations {
		for _, instance := range reservation.Instances {
			instanceID := *instance.InstanceId
			tags := instance.Tags

			var correctedTags []types.Tag
			for _, tag := range tags {
				key := *tag.Key
				value := *tag.Value

				correctedKey := capitalize(key)
				correctedValue := capitalize(value)

				if key != correctedKey || value != correctedValue {
					correctedTags = append(correctedTags, types.Tag{
						Key:   aws.String(correctedKey),
						Value: aws.String(correctedValue),
					})
				}
			}

			if len(correctedTags) > 0 {
				_, err := svc.CreateTags(context.TODO(), &ec2.CreateTagsInput{
					Resources: []string{instanceID},
					Tags:      correctedTags,
				})
				if err != nil {
					log.Printf("Failed to update tags for instance %s: %v", instanceID, err)
				} else {
					fmt.Printf("Corrected tags for instance %s\n", instanceID)
				}
			}
		}
	}

	return map[string]interface{}{
		"statusCode": 200,
		"body":       "Tag correction completed",
	}, nil
}

func capitalize(s string) string {
	if len(s) == 0 {
		return s
	}
	return string(s[0]-32) + s[1:]
}

func main() {
	lambda.Start(handler)
}
